lib/linked_list.rb

```diff
@@ -18,128 +18,269 @@ def initialize
+      return nil if @head.nil?
+      min = @head.data
+      pointer = @head
+      while pointer.next
+        pointer = pointer.next
+        pointer.data < min ? min = pointer.data : min
+      end
+      return min
     end
 
-
     # Additional Exercises 
     # returns the value in the first node
     # returns nil if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1), because the first node is always at the beginning of the lilnked list
+    # Space Complexity: O(1), no extra variable is used in this case
     def get_first
-      raise NotImplementedError
+      return @head.nil? ? nil : @head.data
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), because it needs to loop thru all linked list to add a new node to the end of the list
+    # Space Complexity: O(1), because the variables used here is a constant
     def add_last(value)
-      raise NotImplementedError
+      return @head = Node.new(value) if @head.nil?
+      
+      pointer = @head
+      while pointer.next
+        pointer = pointer.next 
+      end
+      
+      add_node = Node.new(value)
+      pointer.next = add_node
+      return @head
     end
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), because it needs to loop thru all linked list to find out the length of the linked list
+    # Space Complexity: O(1), because the variables used here is a constant
     def length
-      raise NotImplementedError
+      count = 0
+      pointer = @head
+      while pointer
+        count += 1
+        pointer = pointer.next
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), because the worst case scenario is to loop thru all linked list to get the data for input index
+    # Space Complexity: O(1), because the variables used here is a constant
     def get_at_index(index)
-      raise NotImplementedError
+      count = 0
+      pointer = @head
+      while pointer
+        return pointer.data if count == index
+        pointer = pointer.next
+        count += 1
+      end
+      return nil
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n), because it has to loop thru all linked list to print the values
+    # Space Complexity: O(n)/O(1), because add_value depends on the length of the linked list, it would be O(n); 
+    #                         unless we don't need all_value, it would be O(1) for just print out the data.
     def visit

```

Comment:
Your solution would be O(n) for space complexity since you're building an array of all the elements. 

The question is, do you need the array?